### PR TITLE
Wazo 3213 queue interception recording purge

### PR DESCRIPTION
**Pull request by DrPyser**
https://wazo-dev.atlassian.net/browse/WAZO-3213

### CODE REVIEW

**Review by xxbillouxx — DISMISSED** [review]

File: wazo_call_logd/generator.py [hunk 1]
```diff
@@ -154,10 +159,29 @@ def from_cel(self, cels):
...
+            )
+
+            terminated_links = set(
+                cel.linkedid
+                for cel in cels_by_call
+                if cel.eventtype == CELEventType.linkedid_end
+            )
+            unterminated_links = set(
+                linkedid for linkedid in linkedids if linkedid not in terminated_links
+            )
```
Comment (sduthil) [inline wazo_call_logd/generator.py:178]: Suggestion: 
```
unterminated_links = set(linkedids) - terminated_links
```
And maybe check `if len(linkedids) != len(unterminated_links)` first, then compute the difference for logging only.

File: wazo_call_logd/generator.py [hunk 2]
```diff
@@ -186,6 +210,46 @@ def _group_cels_by_linkedid(self, cels):
         cels = sorted(cels, key=attrgetter('linkedid'))
         return groupby(cels, key=attrgetter('linkedid'))
 
+    def _group_cels_by_shared_channels(
```
Comment (sduthil) [inline wazo_call_logd/generator.py:213]: I think this function deserves its own unit tests, what do you think?

**Review by sduthil — CHANGES_REQUESTED** [review]
Very nice :)
Just a few details to fix, and we're good to go!

File: wazo_call_logd/generator.py [hunk 3]
```diff
@@ -186,6 +210,46 @@ def _group_cels_by_linkedid(self, cels):
...
+            for linkedid, cels in groupby(cels, key=attrgetter('linkedid'))
+        ]
+
+        # identify linkedid-based cel sequences that share uniqueids(i.e. channels)
+        # this correlation is transitive, i.e. if a channel is shared between sequence a and b, and between b and c,
+        # then a and c are also correlated
+        correlation_groups: list[(set[str], set[str], list[CEL])] = []
+        for linkedid, cels in linkedid_sequences:
+            uniqueids = set(cel.uniqueid for cel in cels)
+            correlated_sequences = set()  # linkedids of sequences with shared uniqueids
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:228]: True

File: wazo_call_logd/generator.py (hunk 2 above)
Comment (DrPyser) [reply wazo_call_logd/generator.py:212]: Agreed

File: wazo_call_logd/cel_interpretor.py [hunk 4]
```diff
@@ -94,6 +94,55 @@ def _extract_call_log_destination_variables(extra: dict) -> dict:
...
+        groups = match.groupdict()
+        return groups['exten'], groups['context']
+
+
+def _extract_pickup_info(cel: CEL) -> PickupInfo | None:
+    if cel.appname != 'Pickup':
+        return None
+    elif not cel.appdata:
+        logger.error('Missing expected pickup appdata for Pickup application event')
+        return PickupInfo(interceptor_channel=cel.channame)
```
Comment (sduthil) [inline wazo_call_logd/cel_interpretor.py:114]: Is this a valid `PickupInfo`? Is `PickupInfo` consumed somewhere? I could not find where it is used.
Should we return `None` instead?

**Review by sduthil — APPROVED** [review]

**Pull request merged**